Client/WebFilters/DebouncingResourceDownloader.swift

```diff
@@ -80,155 +285,185 @@ public class DebouncingResourceDownloader {
+          // an error presented or anything to happen because iOS is not ready to handle this new rule
+          // But since it's impossible to discriminate a programmer mistake from a new rule added
+          // we at least log this error.
+          log.error("Unsupported rule preference used for debouncing: `\(rawRule)`")
+        } catch {
+          log.error(error)
         }
       }
 
       self.etldToRule = etldToRule
       self.queryToRule = queryToRule
       self.otherRules = otherRules
     }
-
+    
     /// Get redirect url recursively by continually applying the matcher rules to each url returned until we have no more redirects.
-    /// Only handles `http` and `https` schemes.
-    func redirectURLAll(from url: URL) -> URL? {
+    ///
+    /// The following conditions must be met to redirect:
+    /// 1. Must have a `MatcherRule` that satisifes this url
+    /// 2. Extracted URL  not be the same domain as the base URL
+    /// 3. Extracted URL must be a valid URL (Have valid scheme, have an `eTLD+1` or be an IP address etc.. )
+    /// 4. Extracted URL must have a `http` or `https` scheme
+    func redirectChain(from url: URL) -> [(url: URL, rule: RedirectRule)] {
       var redirectingURL = url
-      var result: URL?
+      var result: [(url: URL, rule: RedirectRule)] = []
 
-      while let nextURL = redirectURLOnce(from: redirectingURL) {
-        redirectingURL = nextURL
-        result = nextURL
+      while let nextElement = redirectURLOnce(from: redirectingURL) {
+        redirectingURL = nextElement.url
+        result.append(nextElement)
       }
 
       return result
     }
 
-    /// Get a possible redirect url for the given URL and the given matchers. Will only get the next redirect url.
-    /// Only handles `http` and `https` schemes.
+    /// Get a possible redirect URL and rule for the given URL.
+    /// This only returns the next redirect entry and does not process the returned urls recursively.
     ///
-    /// This code uses the patterns in the given matchers to determine if a redirect action is required.
-    /// If it is, a redirect url will be returned provided we can extract it from the url
+    /// The following conditions must be met to redirect:
+    /// 1. Must have a `MatcherRule` that satisifes this url
+    /// 2. Extracted URL  not be the same domain as the base URL
+    /// 3. Extracted URL must be a valid URL (Have valid scheme, have an `eTLD+1` or be an IP address etc.. )
+    /// 4. Extracted URL must have a `http` or `https` scheme
+    private func redirectURLOnce(from url: URL) -> (url: URL, rule: RedirectRule)? {
+      guard let rule = matchingRedirectRule(for: url) else {
+        return nil
+      }
+      
+      do {
+        guard let extractedURL = try rule.extractRedirectURL(from: url),
+              url.host != extractedURL.host,
+              url.scheme == "http" || url.scheme == "http" else {
```

Comment:
Is this supposed to be "http" and "https"?

Reply:
I fixed this already. But yah https